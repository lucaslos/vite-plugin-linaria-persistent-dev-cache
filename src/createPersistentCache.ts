import type { Result } from '@linaria/babel-preset'
import fs from 'fs'

type FileEntry = {
  code: string
  map: Result['sourceMap']
  cssText: string
  cssSlug: string
}

type Options = {
  cacheFilePath: string
  viteConfigFilePath: string
  packageJsonPath: string
  packageJsonDependencies: string[]
}

export function createPersistentCache({ cacheFilePath }: Options) {
  const cacheFileExists = fs.statSync(cacheFilePath, {
    throwIfNoEntry: false,
  })

  let updateTimeout: NodeJS.Timeout | null = null

  const persistentCache: Record<string, FileEntry> = cacheFileExists
    ? JSON.parse(fs.readFileSync(cacheFilePath, 'utf8'))
    : {}

  function updateCache() {
    if (updateTimeout) {
      clearTimeout(updateTimeout)
    }

    updateTimeout = setTimeout(() => {
      fs.writeFile(cacheFilePath, JSON.stringify(persistentCache), () => {})
    }, 3000)
  }

  function addFile(hash: string, file: FileEntry) {
    persistentCache[hash] = file
    updateCache()
  }

  function getFile(hash: string) {
    return persistentCache[hash]
  }

  function removeFile(fileId: string) {
    // FIX: remove file deps?
    delete persistentCache[hash]
    updateCache()
  }

  return {
    addFile,
    getFile,
    removeFile,
  }
}
