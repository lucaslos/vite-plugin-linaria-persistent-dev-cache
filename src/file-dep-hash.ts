import crypto from "crypto";
import fs from "fs";
import path from "path";

function generateStringHash(str: string) {
  return crypto.createHash("sha1").update(str).digest("hex");
}

type Aliases = {
  find: string;
  replacement: string;
}[];

function getImportsFromCode(code: string, include: RegExp[]) {
  const regex = /from\s+['"]([^'"]+)['"];/g;

  const allPossibleImports = code.matchAll(regex);

  const imports: string[] = [];

  for (const [_, path] of allPossibleImports) {
    if (include.some((pattern) => pattern.test(path))) {
      imports.push(path);
    }
  }

  return imports;
}

const resolveCache = new Map<string, string>();

function getResolvedPath(
  filePath: string,
  aliases: Aliases,
  rootDir: string
): string | false {
  if (resolveCache.has(filePath)) {
    return resolveCache.get(filePath)!;
  }

  let normalizedPath = filePath;

  for (const { find, replacement } of aliases) {
    if (normalizedPath.startsWith(find)) {
      normalizedPath = normalizedPath.replace(find, replacement);
    }
  }

  normalizedPath = path.posix.join(rootDir, normalizedPath);

  const testSuffix = [".tsx", ".ts", "/index.tsx", "/index.ts"];

  for (const suffix of testSuffix) {
    const testURL = `${normalizedPath}${suffix}`;

    if (fs.statSync(testURL, { throwIfNoEntry: false })) {
      return testURL;
    }
  }

  return false;
}

type CodeDependency = {
  fileId: string;
  importPath: string;
  code: string;
};

const codeDepsCache = new Map<
  string,
  { deps: CodeDependency[]; depsId: Set<string> }
>();
const codeDepsCacheFilesIds = new Set<string>();

type Debug = {
  cached: number;
  notCached: number;
  addedToCache: number;
  timing: number;
};

function getAllCodeDeps(
  fileId: string,
  code: string,
  include: RegExp[],
  rootDir: string,
  aliases: Aliases,
  debug: Debug,
  allDepsFileIds: Set<string> = new Set()
): CodeDependency[] {
  if (codeDepsCache.has(fileId)) {
    debug.cached++;
    return codeDepsCache.get(fileId)!.deps;
  }

  const codeImports = getImportsFromCode(code, include);
  const deps = new Map<string, CodeDependency>();

  allDepsFileIds.add(fileId);

  for (const importPath of codeImports) {
    if (deps.has(importPath)) {
      continue;
    }

    const resolvedFiledId = getResolvedPath(importPath, aliases, rootDir);

    if (!resolvedFiledId) {
      continue;
    }

    if (allDepsFileIds.has(resolvedFiledId)) {
      continue;
    }

    allDepsFileIds.add(resolvedFiledId);

    const code = fs.readFileSync(resolvedFiledId, "utf8");

    deps.set(importPath, { fileId: resolvedFiledId, code, importPath });

    const importDeps = getAllCodeDeps(
      resolvedFiledId,
      code,
      include,
      rootDir,
      aliases,
      debug,
      allDepsFileIds
    );

    for (const importResult of importDeps) {
      deps.set(importResult.importPath, importResult);
    }
  }

  const depsArray: CodeDependency[] = [];
  const depsId: Set<string> = new Set();

  for (const dependency of deps.values()) {
    depsId.add(dependency.fileId);
    depsArray.push(dependency);
  }

  // if (codeImports.length === 0) {
  //   debug.addedToCache++;
  //   codeDepsCache.set(fileId, {
  //     deps: [],
  //     depsId: new Set(),
  //   });
  // }

  debug.notCached++;

  return depsArray;
}

function cleanCodeDepsCacheForFile(fileId: string) {
  if (!codeDepsCacheFilesIds.has(fileId)) {
    // TODO: ignore files in the first build
    return;
  }

  for (const [id, cacheEntry] of codeDepsCache.entries()) {
    if (fileId === id) {
      codeDepsCache.delete(id);
      continue;
    }

    if (cacheEntry.depsId.has(fileId)) {
      codeDepsCache.delete(id);
    }
  }
}

export function getCodeHash(
  fileId: string,
  code: string,
  include: RegExp[],
  aliases: Aliases,
  rootDir: string
) {
  // FIX: make debug optional
  const debug: Debug = {
    cached: 0,
    notCached: 0,
    addedToCache: 0,
    timing: 0,
  };

  const start = Date.now();

  const importsMap = getAllCodeDeps(
    fileId,
    code,
    include,
    rootDir,
    aliases,
    debug
  );

  const codeHash = generateStringHash(code);

  let importsHash = "";

  for (const { code, fileId } of importsMap) {
    importsHash += generateStringHash(`${fileId}||${code}`);
  }

  debug.timing = Date.now() - start;

  return {
    hash: `${codeHash}||${generateStringHash(importsHash)}`,
    importsMap,
    debug,
  };
}

export function resetCodeDepsCache() {
  codeDepsCache.clear();
  codeDepsCacheFilesIds.clear();
}
