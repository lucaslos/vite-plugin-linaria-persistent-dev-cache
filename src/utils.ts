import crypto from 'crypto'

export function mapEmplace<K, V>(
  map: Map<K, V>,
  key: K,
  methods: {
    insert?: (key: K, map: Map<K, V>) => V
    update?: (existings: V, key: K, map: Map<K, V>) => V
  },
): V {
  let calledAMethod = false

  if (map.has(key)) {
    if (methods.update) {
      calledAMethod = true
      const newValue = methods.update(map.get(key)!, key, map)

      map.set(key, newValue)
    }
  } else if (methods.insert) {
    calledAMethod = true
    const newValue = methods.insert(key, map)

    map.set(key, newValue)
  }

  return map.get(key)!
}

export function generateStringHash(str: string) {
  return crypto.createHash('sha1').update(str).digest('hex')
}
