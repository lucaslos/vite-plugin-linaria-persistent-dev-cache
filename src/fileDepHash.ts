import fs from 'fs'
import path from 'path'
import { generateStringHash } from './utils'

type Aliases = {
  find: string | RegExp
  replacement: string
}[]

interface FileDepHashConfig {
  include: RegExp[]
  exclude: RegExp[]
  aliases: Aliases
  rootDir: string
  resolveRelative?: boolean
}

interface InstanceProps extends FileDepHashConfig {
  resolveCache: Map<string, string>
  codeDepsCache: Map<string, CacheEntry>
}

function getImportsFromCode(
  code: string,
  { include, exclude }: InstanceProps,
): string[] {
  const codeWithoutComments = code.replace(/\/\/.*|\/\*[^]*?\*\//g, '')

  const regex = /from\s+['"]([^'"]+)['"]/g

  const allPossibleImports = codeWithoutComments.matchAll(regex)

  const imports: string[] = []

  for (const [_, path] of allPossibleImports) {
    if (!include.some((pattern) => pattern.test(path!))) {
      continue
    }

    if (exclude.some((regex) => regex.test(path!))) {
      continue
    }

    imports.push(path!)
  }

  return imports
}

function getResolvedPath(
  filePath: string,
  { aliases, rootDir, resolveCache, resolveRelative }: InstanceProps,
  resolveRelativeFrom: string,
): string | false {
  if (!resolveRelative && resolveCache.has(filePath)) {
    return resolveCache.get(filePath)!
  }

  let normalizedPath = filePath

  if (resolveRelative) {
    normalizedPath = path.resolve(path.dirname(resolveRelativeFrom), filePath)
  } else {
    for (const { find, replacement } of aliases) {
      if (
        typeof find === 'string'
          ? normalizedPath.startsWith(find)
          : find.test(normalizedPath)
      ) {
        normalizedPath = normalizedPath.replace(find, replacement)
      }
    }

    normalizedPath = path.posix.join(rootDir, normalizedPath)
  }

  const fileName = path.basename(normalizedPath)

  const firstLetterIsUpper = fileName[0]?.toUpperCase() === fileName[0]

  const testSuffix = firstLetterIsUpper
    ? ['.tsx', '.ts', '/index.tsx', '/index.ts']
    : ['.ts', '.tsx', '/index.ts', '/index.tsx']

  for (const suffix of testSuffix) {
    const testURL = `${normalizedPath}${suffix}`

    if (fs.statSync(testURL, { throwIfNoEntry: false })) {
      return testURL
    }
  }

  return false
}

function getCodeFromResolvedPath(
  resolvedPath: string,
  config: InstanceProps,
  resolveRelativeFrom: string,
): string {
  let code: string

  try {
    code = fs.readFileSync(resolvedPath, 'utf8')
  } catch (e) {
    for (const [unresolved, resolved] of config.resolveCache) {
      if (resolved === resolvedPath) {
        config.resolveCache.delete(unresolved)

        const resolvedPath = getResolvedPath(
          unresolved,
          config,
          resolveRelativeFrom,
        )

        if (resolvedPath) {
          return getCodeFromResolvedPath(
            resolvedPath,
            config,
            resolveRelativeFrom,
          )
        }
      }
    }

    throw e
  }

  return code
}

type CodeDependency = {
  fileId: string
  code: string
}

type CacheEntry =
  | {
      deps: CodeDependency[]
      depsId: Set<string>
    }
  | false

type Debug = {
  cached: number
  notCached: number
  addedToCache: number
  timing: number
  getAllCodeDepsCalls: number
}

function getEdges(
  code: string,
  parentFileId: string,
  config: InstanceProps,
): string[] {
  const imports = getImportsFromCode(code, config)

  const edges: string[] = []

  for (const importPath of imports) {
    const resolvedFiledId = getResolvedPath(importPath, config, parentFileId)

    if (!resolvedFiledId) {
      continue
    }

    edges.push(resolvedFiledId)
  }

  return edges
}

function mergeEdgeDeps(
  edge: string,
  deps: Map<string, string>,
  config: InstanceProps,
) {
  const cacheEntry = config.codeDepsCache.get(edge)

  if (!cacheEntry) {
    return
  }

  const edgeDeps = cacheEntry.deps

  for (const dep of edgeDeps) {
    deps.set(dep.fileId, dep.code)
  }
}

function getAllCodeDeps(
  fileId: string,
  code: string,
  config: InstanceProps,
  debug: Debug,
  visited: Map<string, string> = new Map(),
  inPath: Set<string> = new Set(),
  deepth = 0,
): { deps: CodeDependency[]; hasCircularDep: boolean } {
  debug.getAllCodeDepsCalls++

  const cachedValue = config.codeDepsCache.get(fileId)

  if (cachedValue) {
    debug.cached++

    return { deps: cachedValue.deps, hasCircularDep: false }
  }

  visited.set(fileId, code)
  inPath.add(fileId)

  const edges = getEdges(code, fileId, config)

  const deps = new Map<string, string>()

  let hasCircularDep = false

  for (const edge of edges) {
    if (inPath.has(edge)) {
      hasCircularDep = true
      continue
    }

    let edgeHasCircularDep = false

    let edgeCode: string

    if (!visited.has(edge)) {
      edgeCode = getCodeFromResolvedPath(edge, config, fileId)

      edgeHasCircularDep = getAllCodeDeps(
        edge,
        edgeCode,
        config,
        debug,
        visited,
        inPath,
        deepth + 1,
      ).hasCircularDep
    } else {
      edgeHasCircularDep = !config.codeDepsCache.get(edge)

      edgeCode = visited.get(edge)!
    }

    if (edgeHasCircularDep) {
      hasCircularDep = true
      continue
    }

    deps.set(edge, edgeCode)

    mergeEdgeDeps(edge, deps, config)
  }

  inPath.delete(fileId)

  if (deepth === 0) {
    visited.delete(fileId)
  }

  const depsArray: CodeDependency[] = []
  const depsId = new Set<string>()

  if (hasCircularDep) {
    if (deepth === 0) {
      populateDepsArray(visited, depsArray, depsId)

      debug.addedToCache++
      config.codeDepsCache.set(fileId, {
        deps: depsArray,
        depsId,
      })
    }
    //
    else {
      debug.notCached
      config.codeDepsCache.set(fileId, false)
    }
  }
  //
  else {
    populateDepsArray(deps, depsArray, depsId)

    debug.addedToCache++
    config.codeDepsCache.set(fileId, {
      deps: depsArray,
      depsId,
    })
  }

  return { deps: depsArray, hasCircularDep }
}

function populateDepsArray(
  deps: Map<string, string>,
  depsArray: CodeDependency[],
  depsId: Set<string>,
) {
  for (const [depFileId, code] of deps) {
    depsArray.push({
      fileId: depFileId,
      code: code,
    })

    depsId.add(depFileId)
  }
}

function cleanCodeDepsCacheForFile(
  fileId: string,
  { codeDepsCache }: InstanceProps,
) {
  for (const [id, cacheEntry] of codeDepsCache.entries()) {
    if (fileId === id) {
      codeDepsCache.delete(id)
      continue
    }

    if (cacheEntry && cacheEntry.depsId.has(fileId)) {
      codeDepsCache.delete(id)
    }
  }
}

type CodeHashResult = {
  hash: string
  importsMap: CodeDependency[]
  debug: Debug
}

function getCodeHash(
  fileId: string,
  code: string,
  config: InstanceProps,
): CodeHashResult {
  const debug: Debug = {
    cached: 0,
    notCached: 0,
    addedToCache: 0,
    timing: 0,
    getAllCodeDepsCalls: 0,
  }

  const start = Date.now()

  const importsMap = getAllCodeDeps(fileId, code, config, debug)

  const codeHash = generateStringHash(code)

  let importsHash = ''

  for (const { code, fileId } of importsMap.deps) {
    importsHash += generateStringHash(`${fileId}||${code}`)
  }

  debug.timing = Date.now() - start

  return {
    hash: `${codeHash}||${generateStringHash(importsHash)}`,
    importsMap: importsMap.deps,
    debug,
  }
}

export type FileDepHashInstance = {
  resetCache: () => void
  getHash: (fileId: string, code: string) => CodeHashResult
  getCodeDepsCache: () => Map<string, CacheEntry>
  cleanCacheForFile: (fileId: string) => void
}

export function createFileDepHash(
  config: FileDepHashConfig,
): FileDepHashInstance {
  const resolveCache = new Map<string, string>()
  const codeDepsCache = new Map<string, CacheEntry>()
  const changedAfterInitialBuild = new Set<string>()

  const instance: InstanceProps = {
    ...config,
    resolveCache,
    codeDepsCache,
  }

  function resetCache() {
    codeDepsCache.clear()
    changedAfterInitialBuild.clear()
  }

  function getCodeDepsCache() {
    return codeDepsCache
  }

  function cleanCacheForFile(fileId: string) {
    cleanCodeDepsCacheForFile(fileId, instance)
  }

  function getHash(fileId: string, code: string): CodeHashResult {
    return getCodeHash(fileId, code, instance)
  }

  return {
    resetCache,
    getHash,
    getCodeDepsCache,
    cleanCacheForFile,
  }
}

export const testOnly = {
  getImportsFromCode,
}
