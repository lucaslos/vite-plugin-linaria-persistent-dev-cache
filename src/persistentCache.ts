import type { Result } from '@linaria/babel-preset'
import fs from 'fs'
import path from 'path'
import { createFileDepHash } from './fileDepHash'
import { generateStringHash } from './utils'

const cacheFormatVersion = 1

interface FileEntry {
  code: string
  map: Result['sourceMap']
  cssText: string
  cssSlug: string
}

interface CacheEntry extends FileEntry {
  timestamp: number
  fileId: string
}

interface CacheFile {
  version: number
  rootDir: string
  lockFileHash: string
  viteConfigHash: string
  results: Record<string, CacheEntry>
  fileHashes: Record<string, string>
}

type Options = {
  cacheFilePath: string
  viteConfigFilePath: string
  lockFilePath: string
  rootDir: string
  debug?: boolean
  _readFile?: (path: string) => string | null
  _writeFile?: (path: string, data: string) => void
  _writeDebounce?: number
  _getNow?: () => number
}

function defaultReadFile(filePath: string): string | null {
  const fileExists = fs.statSync(filePath, {
    throwIfNoEntry: false,
  })

  if (!fileExists) {
    return null
  }

  return fs.readFileSync(filePath, 'utf8')
}

function defaultWriteFile(filePath: string, data: string) {
  const dirname = path.dirname(filePath)

  if (!fs.existsSync(dirname)) {
    fs.mkdirSync(dirname, { recursive: true })
  }

  fs.writeFile(filePath, data, (err) => {
    if (err) {
      throw err
    }
  })
}

export function createPersistentCache({
  cacheFilePath,
  _readFile = defaultReadFile,
  _writeFile = defaultWriteFile,
  _writeDebounce = 3000,
  lockFilePath,
  rootDir,
  debug,
  viteConfigFilePath,
  _getNow = Date.now,
}: Options) {
  const cacheFileJSON = _readFile(cacheFilePath)

  let updateTimeout: NodeJS.Timeout | null = null

  let persistentCache: CacheFile

  if (cacheFileJSON) {
    persistentCache = JSON.parse(cacheFileJSON)
  } else {
    persistentCache = {
      version: cacheFormatVersion,
      lockFileHash: getLockFileHash(),
      viteConfigHash: getViteConfigHash(),
      results: {},
      fileHashes: {},
      rootDir,
    }
  }

  function debugLog(...args: any[]) {
    if (debug) {
      console.log('[linaria]', ...args)
    }
  }

  function checkConfigFiles() {
    const lockFileHash = getLockFileHash()
    const viteConfigHash = getViteConfigHash()

    debugLog('checking config files')

    if (
      persistentCache.version !== cacheFormatVersion ||
      lockFileHash !== persistentCache.lockFileHash ||
      viteConfigHash !== persistentCache.viteConfigHash
    ) {
      debugLog('cache reseted')

      persistentCache = {
        version: cacheFormatVersion,
        lockFileHash,
        viteConfigHash,
        results: {},
        fileHashes: {},
        rootDir,
      }

      updateCache()
    }
  }

  function getLockFileHash() {
    const resolvedPath = lockFilePath

    const lockFileContent = _readFile(lockFilePath)

    if (!lockFileContent) {
      throw new Error(`Could not read lock file at ${resolvedPath}`)
    }

    return generateStringHash(lockFileContent)
  }

  function getViteConfigHash() {
    const viteConfigContent = _readFile(viteConfigFilePath)

    if (!viteConfigContent) {
      throw new Error(
        `Could not read vite config file at ${viteConfigFilePath}`,
      )
    }

    const fileDepHash = createFileDepHash({
      rootDir,
      aliases: [],
      resolveRelative: true,
      include: [/^\.+/],
      exclude: [],
    })

    return fileDepHash.getHash(viteConfigFilePath, viteConfigContent).hash
  }

  function cacheEntryIsExpired(entry: CacheEntry, now = _getNow()) {
    const timeSinceLastUpdate = now - entry.timestamp

    return timeSinceLastUpdate > 24 * 7 * 60 * 60 * 1000
  }

  function deleteEntry(hash: string, entry: CacheEntry) {
    const fileId = entry.fileId

    delete persistentCache.results[hash]
    delete persistentCache.fileHashes[getFileId(fileId)]
  }

  function deleteOldFiles() {
    const now = _getNow()

    Object.entries(persistentCache.results).forEach(([hash, entry]) => {
      if (cacheEntryIsExpired(entry, now)) {
        deleteEntry(hash, entry)
      }
    })
  }

  function updateCache() {
    if (updateTimeout) {
      clearTimeout(updateTimeout)
    }

    updateTimeout = setTimeout(() => {
      deleteOldFiles()

      _writeFile(cacheFilePath, JSON.stringify(persistentCache))
    }, _writeDebounce)
  }

  function compressFileId(fileId: string) {
    return fileId.replace(rootDir, '')
  }

  function getFileId(filePath: string) {
    return `${rootDir}${filePath}`
  }

  function addFile(hash: string, fileId: string, result: FileEntry) {
    const fileIdCompressed = compressFileId(fileId)

    persistentCache.results[hash] = {
      ...result,
      fileId: fileIdCompressed,
      timestamp: _getNow(),
    }

    const previousFileHash = persistentCache.fileHashes[fileIdCompressed]

    if (previousFileHash) {
      delete persistentCache.results[previousFileHash]
    }

    persistentCache.fileHashes[fileIdCompressed] = hash
    updateCache()
  }

  function getFile(hash: string) {
    const result = persistentCache.results[hash]

    if (!result) return null

    if (cacheEntryIsExpired(result)) {
      deleteEntry(hash, result)
      return null
    }

    return result
  }

  return {
    addFile,
    getFile,
    checkConfigFiles,
  }
}
