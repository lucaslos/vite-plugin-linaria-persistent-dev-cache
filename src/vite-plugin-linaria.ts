/**
 * This file contains a Rollup loader for Linaria.
 * It uses the transform.ts function to generate class names from source code,
 * returns transformed code without template literals and attaches generated source maps
 */
import { EvalCache, Module, slugify, transform } from '@linaria/babel-preset'
import fs from 'fs'
import path from 'path'
import { normalizePath, Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import { createFileDepHash, FileDepHashInstance } from './fileDepHash'
import { createPersistentCache } from './persistentCache'

type RollupPluginOptions = {
  sourceMap?: boolean
  persistentCachePath?: string
  disableDevPersistentCache?: boolean
  include: RegExp[]
  exclude?: RegExp[]
  lockFilePath: string
  viteConfigFilePath?: string
  debug?: boolean
}

export default function linaria({
  sourceMap,
  persistentCachePath = './node_modules/.linaria-cache/cache.json',
  disableDevPersistentCache,
  include,
  exclude = [],
  viteConfigFilePath = './vite.config.ts',
  lockFilePath,
  debug,
}: RollupPluginOptions): Plugin {
  const root = normalizePath(process.cwd())
  let config: ResolvedConfig
  let server: ViteDevServer | undefined

  const virtualCssFiles = new Map<string, string>()

  const lockFileAbsPath = normalizePath(path.resolve(root, lockFilePath))

  const persistentCache = createPersistentCache({
    cacheFilePath: persistentCachePath,
    viteConfigFilePath: path.resolve(root, viteConfigFilePath),
    lockFilePath: lockFileAbsPath,
    rootDir: root,
    debug: false,
  })

  let fileDepHash: FileDepHashInstance

  function getVirtualName(id: string) {
    return `@linaria-css-cache/${slugify(id)}.css`
  }

  type StatsMode = 'create' | 'cached' | 'skiped'

  let logStats:
    | ((startTime: number, mode: StatsMode, file: string) => void)
    | null = null

  if (debug) {
    const stats: Record<
      `${StatsMode}Time` | `${StatsMode}Count` | 'total',
      number
    > = {
      total: 0,
      createTime: 0,
      cachedTime: 0,
      skipedTime: 0,
      createCount: 0,
      cachedCount: 0,
      skipedCount: 0,
    }

    const uncachedFiles: string[] = []
    const skipedFiles: string[] = []

    let statsTimeout: any
    logStats = (startTime, mode, file: string) => {
      const delta = Date.now() - startTime
      stats.total += delta
      stats[`${mode}Time`] += delta
      stats[`${mode}Count`]++

      if (mode === 'create') {
        uncachedFiles.push(file)
      }

      if (mode === 'skiped') {
        skipedFiles.push(file)
      }

      clearTimeout(statsTimeout)
      statsTimeout = setTimeout(() => {
        console.log(stats, file)

        fs.writeFileSync(
          'linaria-plugin-stats.json',
          JSON.stringify(stats, null, 2),
        )
      }, 2000)
    }
  }

  return {
    name: 'linaria',
    configResolved(resolvedConfig) {
      config = resolvedConfig

      fileDepHash = createFileDepHash({
        rootDir: root,
        aliases: config.resolve.alias,
        include,
        exclude,
      })
    },
    configureServer(_server) {
      server = _server
    },
    load(id: string) {
      return virtualCssFiles.get(id)
    },
    resolveId(importee: string) {
      if (virtualCssFiles.has(importee)) return importee

      if (importee.startsWith('/linaria-css-cache/')) {
        return importee.replace('/linaria-css-cache/', '@linaria-css-cache/')
      }

      return
    },
    transform(code: string, id: string) {
      if (
        id.includes('node_modules') ||
        virtualCssFiles.has(id) ||
        !code.includes('@linaria/')
      ) {
        return
      }

      const startTime = Date.now()

      const isDevMode = config.command === 'serve'
      const enablePersistentCache = isDevMode && !disableDevPersistentCache

      let hash: string | false = false

      if (enablePersistentCache) {
        hash = fileDepHash.getHash(id, code).hash

        const cached = hash && persistentCache.getFile(hash)

        if (cached) {
          const virtualName = getVirtualName(id)

          virtualCssFiles.set(virtualName, cached.cssText)

          logStats?.(startTime, 'cached', id)

          return cached
        }
      }

      const originalResolver = Module._resolveFilename

      Module._resolveFilename = aliasResolver(config, originalResolver, root)

      const result = transform(code, {
        filename: id,
        pluginOptions: {
          displayName: isDevMode,
        },
      })

      Module._resolveFilename = originalResolver

      if (!result.cssText) {
        logStats?.(startTime, 'skiped', id)

        return
      }

      let { cssText } = result

      const virtualName = getVirtualName(id)

      if (sourceMap && result.cssSourceMapText) {
        const map = Buffer.from(result.cssSourceMapText).toString('base64')
        cssText += `/*# sourceMappingURL=data:application/json;base64,${map}*/`
      }

      if (server) {
        const module = server.moduleGraph.getModuleById(virtualName)

        if (module) {
          server.moduleGraph.invalidateModule(module)

          console.log(module.id)

          server.ws.send({
            type: 'update',
            updates: [
              {
                acceptedPath: module.id!,
                path: module.id!,
                timestamp: Date.now(),
                type: 'js-update',
              },
            ],
          })
        }
      }

      virtualCssFiles.set(virtualName, cssText)

      result.code += `\nimport "${virtualName}";\n`

      if (hash) {
        persistentCache.addFile(hash, id, {
          code: result.code,
          cssText,
          map: result.sourceMap,
        })
      }

      logStats?.(startTime, 'create', id)

      return { code: result.code, map: result.sourceMap }
    },
    handleHotUpdate({ file }) {
      fileDepHash.cleanCacheForFile(file)
      // FIX: eval cache is working?
      EvalCache.clearForFile(file)

      if (file === lockFileAbsPath) {
        persistentCache.checkConfigFiles()
      }
    },
    buildStart() {
      persistentCache.checkConfigFiles()
    },
  }
}

type ResolveFilename = (
  id: string,
  options: { id: string; filename: string; paths: string[] },
) => string

function aliasResolver(
  config: ResolvedConfig,
  originalResolveFilename: ResolveFilename,
  root: string,
): ResolveFilename {
  return (id, options) => {
    let aliasedPath: string | undefined = undefined

    for (const { find, replacement } of config.resolve.alias) {
      const matches =
        typeof find === 'string' ? id.startsWith(find) : find.test(id)

      if (matches) {
        aliasedPath = id.replace(find, replacement)
        break
      }
    }

    const finalPath = aliasedPath
      ? originalResolveFilename(path.join(root, aliasedPath), options)
      : originalResolveFilename(id, options)

    return finalPath
  }
}
