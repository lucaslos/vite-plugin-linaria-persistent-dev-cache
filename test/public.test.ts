import { beforeEach, describe, expect, test } from "vitest";
import { resetCodeDepsCache } from "../src/file-dep-hash";
import { getPublicFileDepHash } from "./utils/setup";

beforeEach(() => {
  resetCodeDepsCache();
});

test("simple case", () => {
  const result = getPublicFileDepHash("./src/base/base.tsx");

  expect(result.importsMap.length).toEqual(2);
  expect(result.importsMap.map((item) => item.fileId)).toMatchSnapshot();
});

test("circular dependency", () => {
  const result = getPublicFileDepHash("./src/circular/circular.tsx");

  expect(result.importsMap.length).toEqual(2);
  expect(result.importsMap.map((item) => item.fileId)).toMatchSnapshot();
})
