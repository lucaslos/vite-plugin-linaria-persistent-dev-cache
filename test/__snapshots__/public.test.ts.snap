// Vitest Snapshot v1

exports[`circular dependency 1`] = `
[
  "C:/Users/lucas/Github/file-dep-hash/test/___mocks___/public/src/circular/dep1.ts",
  "C:/Users/lucas/Github/file-dep-hash/test/___mocks___/public/src/circular/dep2.ts",
]
`;

exports[`simple case 1`] = `
[
  "C:/Users/lucas/Github/file-dep-hash/test/___mocks___/public/src/base/dep1.ts",
  "C:/Users/lucas/Github/file-dep-hash/test/___mocks___/public/src/base/dep2.ts",
]
`;
